### PR TITLE
[AutoDiff] TF-308: Fix reabstraction thunk logic.

### PR DESCRIPTION
**Pull request by dan-zheng**
Make result type reabstraction logic robust during differentiation thunk
generation. The previous implementation was not robust for all abstraction
patterns. Add test.

Resolves [TF-308](https://bugs.swift.org/browse/TF-308).

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -2164,33 +2164,29 @@ static SILFunction *getOrCreateReabstractionThunk(SILOptFunctionBuilder &fb,
...
-  for (unsigned resIdx : range(toType->getResults().size())) {
+  // Handle indirect results.
+  assert(fromType->getNumResults() == toType->getNumResults());
+  for (unsigned resIdx : range(toType->getNumResults())) {
     auto fromRes = fromConv.getResults()[resIdx];
     auto toRes = toConv.getResults()[resIdx];
     // No abstraction mismatch.
-    if (fromRes.isFormalIndirect() == toRes.isFormalIndirect())
+    if (fromRes.isFormalIndirect() == toRes.isFormalIndirect()) {
+      // If result typess are indirect, directly pass as next argument.
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
      // If result types are indirect, directly pass as next argument.
```

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks for the catch. Fixed in https://github.com/apple/swift/pull/22944/commits/c3beaef04711829186344f6c27bf43072e673412.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**